src/cargo/sources/registry/mod.rs

```diff
@@ -638,19 +645,37 @@ impl<'cfg> Source for RegistrySource<'cfg> {
     }
 
     fn download(&mut self, package: PackageId) -> CargoResult<MaybePackage> {
-        let hash = self.index.hash(package, &mut *self.ops)?;
-        match self.ops.download(package, hash)? {
-            MaybeLock::Ready(file) => self.get_pkg(package, &file).map(MaybePackage::Ready),
-            MaybeLock::Download { url, descriptor } => {
-                Ok(MaybePackage::Download { url, descriptor })
+        loop {
+            match self.index.hash(package, &mut *self.ops)? {
+                Poll::Ready(hash) => {
+                    return match self.ops.download(package, hash)? {
+                        MaybeLock::Ready(file) => {
+                            self.get_pkg(package, &file).map(MaybePackage::Ready)
+                        }
+                        MaybeLock::Download { url, descriptor } => {
+                            Ok(MaybePackage::Download { url, descriptor })
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    // TODO: dont hot loop for it to be Ready
```

Comment:
Same here: also return `Poll`?

Reply:
Or actually I would say that we can assert this isn't pending, since by this point if we're downloading crates the index entry for the crate should always be available.